### PR TITLE
Product Editor: Handle loading of product editor when global settings are not available on initial render

### LINKED ISSUE #45461: Product Editor: Loading of editor can fail depending on the order of web page script loading and execution
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

The product editor depends on the `productBlockEditorSettings` global, which is on the `window` global object. This global is declared and set with an inline script ([src](https://github.com/woocommerce/woocommerce/blob/7fe00de7278cbc0d7806e1733b8df7320feef933/plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php#L100)), and referenced in the Editor functional component ([src](https://github.com/woocommerce/woocommerce/blob/7fe00de7278cbc0d7806e1733b8df7320feef933/plugins/woocommerce-admin/client/products/product-page.tsx#L105)). 

There are a few things that should be fixed to make this more robust:

- Reference the global as `window.productBlockEditorSettings` - this will prevent the `ReferenceError` and instead return `undefined`
- Modify the `Editor` to wait until the `window.productBlockEditorSettings` global is available before rendering the form... stay in the loading state until then.

### Expected behavior

The product editor loads correctly 100% of the time and is not susceptible to differences due to browser script loading and execution differences (which can manifest themselves depending on what other scripts are loaded on the page).

### Actual behavior

There are scenarios where the product editor fails to load.

One such scenario has been observed on a Woo Express site.

https://github.com/woocommerce/woocommerce/assets/2098816/5bfd0fd9-5f12-4763-8e9a-2d9080112685

image (2)

### Steps to reproduce

I do not yet have a 100% reproducible set of steps that will always cause this failure.

- JN
    - with Beta Tester and WooCommerce 8.6.1 or 8.7.0.rc.1 - can’t reproduce
- Woo Express
    - with Beta Tester and WooCommerce 8.7.0.11 (managed) - reproduced
    - with Beta Tester and WooCommerce 8.6.1 (unmanaged) - can’t reproduce
    - with Beta Tester and WooCommerce 8.7.0.rc.1 - reproduced once; then couldn’t again (even after going back and forth between 8.6.1 and 8.7.rc.1 many times)

So, this definitely feels like a timing issue; I’m pretty confident that given enough time, somebody would be able to reproduce it on 8.6.1 or earlier even; and, in 8.7.0, it is sporadic.

**On an installation where this is reproducible, there is also a workaround that doesn’t involve disabling any other plugins…**

- Go to **WooCommerce** > **Home**
- Reload page (if page didn’t load correctly, which it won’t if the product editor didn’t load)
- Go to **Products** > **Add New**

### WordPress Environment

n/a

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by mattsherman**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR moves the accessing of the global `window.productBlockEditorSettings` object from the product pages to the `BlockEditor` component itself, and also puts in place handling to account for timing issues where the global may not be set when the `BlockEditor` is initially rendered.

As mentioned in #45461, we have seen intermittent issues on Woo Express sites where the product editor would fail to load because of this. Sometimes, simply refreshing the page would fix the issue. It is believed that this issue is *not* specific to Woo Express, and this may have even been seen locally in a dev environment once before this issue was explicitly known.

Closes #45461.

### How to test the changes in this Pull Request:

With a WooCommerce dev env with the new product editor enabled (**WooCommerce** > **Settings** > **Advanced** > **Features** > **Experimental features**)...

1. Go to **Products** > **Add New**
2. Verify the product editor loads.
3. Refresh the page.
4. Verify the product editor loads.
5. Try refreshing again and again.
6. Try activating/deactivating additional plugins, especially the WooCommerce Beta Tester, to try and see if you can trigger the loading failure. Unfortunately, this is a timing issue and is not 100% reproducible, so it is difficult to verify with 100% certainty that it is fixed.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/components/block-editor/block-editor.tsx [hunk 1]
```diff
@@ -91,7 +91,35 @@ export function BlockEditor( {
...
+				timeoutId = setTimeout( checkSettingsGlobal, 100 );
+			}
+		};
+
+		checkSettingsGlobal();
+
+		return () => {
+			clearTimeout( timeoutId );
+		};
+	}, [] );
```
Comment (retrofox) [inline packages/js/product-editor/src/components/block-editor/block-editor.tsx:114]: This way to ensure loading the global settings looks interesting.  Do you know if there is an endpoint to request the setting?

Comment (mattsherman) [reply packages/js/product-editor/src/components/block-editor/block-editor.tsx:114]: There is no endpoint currently. Certainly an endpoint would be a good future step. Some of the settings we put in the endpoint have been considered for separate endpoints (such as the product templates).

This is a minimal change to address the immediate issue.

**Review by retrofox — APPROVED** [review]
LGTM 🚀 !

File: packages/js/product-editor/src/components/block-editor/block-editor.tsx [hunk 2]
```diff
@@ -91,7 +91,35 @@ export function BlockEditor( {
 		return () => window.removeEventListener( 'scroll', wpPinMenuEvent );
 	}, [] );
 
-	const settings = useMemo< Partial< ProductEditorSettings > >( () => {
+	const [ settingsGlobal, setSettingsGlobal ] = useState<
+		Partial< ProductEditorSettings > | undefined
+	>( undefined );
+
```
Comment (retrofox) [inline packages/js/product-editor/src/components/block-editor/block-editor.tsx:97]: Adding a comment about it won't hurt. Anyway let's do it in a follow up

**Pull request merged**